code/android-lifecycle-library/src/androidTest/java/com/adobe/marketing/mobile/lifecycle/LifecycleFunctionalTest.java

```diff
@@ -0,0 +1,552 @@
+		assertEquals(2, mockExtensionApi.createdSharedState.size());
+
+		assertEquals(TimeUnit.MILLISECONDS.toSeconds(firstSessionStartTimeMillis), mockExtensionApi.createdSharedState.get(0).get(SESSION_START_TIMESTAMP));
+		assertEquals(TimeUnit.MILLISECONDS.toSeconds(firstSessionStartTimeMillis) + 10, mockExtensionApi.createdSharedState.get(1).get(SESSION_START_TIMESTAMP));
+	}
+
+	@Test
+	public void testLifecycle__When__Start__Then__DispatchLifecycleContextDataUpdate_SecondLaunchEvent() {
+		// setup
+		long firstSessionStartTimeMillis = currentTimestampMillis;
+		long firstSessionPauseTimeMillis = firstSessionStartTimeMillis + 100L;
+		long secondSessionStartTimeMillis = firstSessionStartTimeMillis + TimeUnit.SECONDS.toMillis(40);
+
+		// test
+		mockExtensionApi.simulateComingEvent(createStartEvent(null, firstSessionStartTimeMillis));
+		mockExtensionApi.simulateComingEvent(createPauseEvent(firstSessionPauseTimeMillis));
+		mockExtensionApi.simulateComingEvent(createStartEvent(null, secondSessionStartTimeMillis));
+
+		// verify
+		assertEquals(2, mockExtensionApi.dispatchedEvents.size());
+		assertEquals(2, mockExtensionApi.createdSharedState.size());
+
+		Map<String, String> expectedContextData = new HashMap<String, String>() {
+			{
+				put(LAUNCH_EVENT, "LaunchEvent");
+				put(HOUR_OF_DAY, hourOfDay);
+				put(DAY_OF_WEEK, dayOfWeek);
+				put(LAUNCHES, "2");
+				put(PREVIOUS_OS, "TEST_OS 5.55");
+				put(OPERATING_SYSTEM, "TEST_OS 5.55");
+				put(LOCALE, "en-US");
+				put(DEVICE_RESOLUTION, "100x100");
+				put(CARRIER_NAME, "TEST_CARRIER");
+				put(DEVICE_NAME, "deviceName");
+				put(PREVIOUS_APPID, "TEST_APPLICATION_NAME 1.1 (12345)");
+				put(APP_ID, "TEST_APPLICATION_NAME 1.1 (12345)");
+				put(RUN_MODE, "APPLICATION");
+				put(DAYS_SINCE_FIRST_LAUNCH, "0");
+				put(DAYS_SINCE_LAST_LAUNCH, "0");
+				put(IGNORED_SESSION_LENGTH, "0");
+			}
+		};
+		assertEquals(expectedContextData, mockExtensionApi.dispatchedEvents.get(1).getEventData().get(LIFECYCLE_CONTEXT_DATA));
+		assertEquals(expectedContextData, mockExtensionApi.createdSharedState.get(1).get(LIFECYCLE_CONTEXT_DATA));
+	}
+
+	@Test
+	public void testLifecycle__When__SecondLaunch_AfterSessionTimeout__Then__GetLaunchEvent() {
+		// setup
+		long firstSessionStartTimeMillis = currentTimestampMillis;
+		long firstSessionPauseTimeMillis = firstSessionStartTimeMillis +  TimeUnit.SECONDS.toMillis(10);
+		long secondSessionStartTimeMillis = firstSessionStartTimeMillis + TimeUnit.SECONDS.toMillis(40);
+
+		// test
+		mockExtensionApi.simulateComingEvent(createStartEvent(null, firstSessionStartTimeMillis));
+		mockExtensionApi.simulateComingEvent(createPauseEvent(firstSessionPauseTimeMillis));
+		mockExtensionApi.simulateComingEvent(createStartEvent(null, secondSessionStartTimeMillis));
+
+		// verify
+		assertEquals(2, mockExtensionApi.dispatchedEvents.size());
+		assertEquals(2, mockExtensionApi.createdSharedState.size());
```

Comment:
nit: Can you add a comment on what are the expected events? 
IMO it will be easier to follow if we read the event to variable with descriptive names and assert it.
val firstSessionStartEvent = ... 
// assert the event data. 

val firstSessionSharedState = ... 
// assert the shared state

Reply:
nit: When asserting event data, i would recommend constructing the entire dictionary and asserting it. It would improve the readability of tests. 

assertEquals(expectedData, event.getEventData())